### ISSUE TITLE
add an event for "TLS handshake completed"

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6024]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#6024|
|Type|enhancement|
|Created|2012-09-28 01:20:04Z|
|Branch|https://github.com/twisted/twisted/tree/handshake-6024-2|

As per [#3687](https://github.com/twisted/twisted/issues/3687), it would be useful to know when TLS negotiation is completed so that application-level decisions which need to be made at the beginning of the connection (or the SSL connection, for `startTLS`) but take certificate information, SNI or other TLS parameters into account.

Attachments:

* [tls_notify.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/663/663e63a44092e9020491a7c92832473191208cf2/6144d0be695152e48ab857bcf60aae3f207aa631.patch) (15286 bytes) - added by amluto on 2013-10-11 18:11:31Z - tls-notify-handshake-done-6204-1.patch
* [tls-when-handshake-done-6204-2-reviewfixes.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/663/663e63a44092e9020491a7c92832473191208cf2/fa6fd3cde8d1931239c70472ddd3df360987d1d5.patch) (7339 bytes) - added by amluto on 2013-11-11 23:42:00Z - follow-up patch to address some review comments
* [tls-when-handshake-done-6204-3-clean-up-tests.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/663/663e63a44092e9020491a7c92832473191208cf2/558ad9a6c8e106f2dce7c9c17c0ece7cb80d8bba.patch) (8606 bytes) - added by amluto on 2013-11-11 23:46:13Z - follow-up patch to remove HandshakeCallbackContextFactory
* [tls_notify_v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/663/663e63a44092e9020491a7c92832473191208cf2/9c6d66131b41f2c576ec5de7a05988fc7e544df8.patch) (21100 bytes) - added by amluto on 2013-11-13 17:47:41Z - Version 3, against trunk
* [tls_notify_6204_v4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/663/663e63a44092e9020491a7c92832473191208cf2/10915d36da2e97df39fab906ca4da93e31907ec1.patch) (10659 bytes) - added by amluto on 2013-11-15 23:21:34Z - Further fixes -- applies on top of svn://svn.twistedmatrix.com/svn/Twisted/branches/tls-handshake-notification-6024@40726 bbbe8e31-12d6-0310-92fd-ac37d47ddeeb
* [tls_notify_6204_v4_startTLS.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/663/663e63a44092e9020491a7c92832473191208cf2/209f6a6e6ad01db66a518d9deb99457dc124a4f7.patch) (499 bytes) - added by shawnn on 2013-11-30 02:40:58Z - Added support for whenHandshakeDone when using startTLS, applies on top of amluto's tls_notify_6204_v4.patch

<details><summary>Searchable metadata</summary>

```
trac-id__6024 6024
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__handshake_6024_2 handshake-6024-2
branch_author__exarkun__glyph__habnabit exarkun, glyph, habnabit
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1348795204000000 1348795204000000
changetime__1467973441267739 1467973441267739
version__None None
owner__glyph glyph
cc__amluto cc__habnabit cc__adiroiban cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Implementation thought: we should do this with an interface, in the style of `IHalfCloseableProtocol`, that gets invoked if it's provided.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

[#6769](https://github.com/twisted/twisted/issues/6769) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|amluto commented|
|-|-|

I think that we shouldn't be calling connectionMade until the handshake is done.  Maybe this should be selectable.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|amluto commented|
|-|-|

Here's a candidate fix (which could also be considered a fix for [#6782](https://github.com/twisted/twisted/issues/6782)).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_6"></a>|@lvh set owner to amluto|
|-|-|

Hi! Thanks for working on this, amluto!

A few nitpicks:

 1. Please document the types of the @ivars
 2. Please don't put empty lines in between ivar defs
 1. ```_handshakeDeferreds``` stops being a list when the handshake is done, but the ivar docs don't specify that
 3. The following thing is removed:

```
247	 	        <SNIP SNIP> This 
248	 	        is used to control error reporting behavior.  If the handshake has not 
249	 	        completed, the underlying L{OpenSSL.SSL.Error} will be passed to the 
250	 	        application's C{connectionLost} method.  If it has completed, any 
251	 	        unexpected L{OpenSSL.SSL.Error} will be turned into a 
252	 	        L{ConnectionLost}.  This is weird; however, it is simply an attempt at 
253	 	        a faithful re-implementation of the behavior provided by 
254	 	        L{twisted.internet.ssl}. 
```

Why? AFAICT it's still true (I'll happily admit it's noisy, but it should probably remain documented).

 4. ```def notifyHandshakeDone(self):``` is missing a docstring
 5. ```def __tryHandshake(self):``` please use a single underscore for private methods
 6. What happens to the deferred previously returned from ```HandshakeCallbackContextFactory```?

Thanks again, and I look forward to merging this patch :)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

I would suggest renaming the transport method `whenHandhakeDone`.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|amluto commented|
|-|-|

Replying to [lvh](#note_6):
> Hi! Thanks for working on this, amluto!
> 
> A few nitpicks:
> 
>  1. Please document the types of the @ivars

Will do.

>  2. Please don't put empty lines in between ivar defs

Will do.

>  3. ```_handshakeDeferreds``` stops being a list when the handshake is done, but the ivar docs don't specify that

Will do.

>  4. The following thing is removed:
> 
> ```
> 247	 	        <SNIP SNIP> This 
> 248	 	        is used to control error reporting behavior.  If the handshake has not 
> 249	 	        completed, the underlying L{OpenSSL.SSL.Error} will be passed to the 
> 250	 	        application's C{connectionLost} method.  If it has completed, any 
> 251	 	        unexpected L{OpenSSL.SSL.Error} will be turned into a 
> 252	 	        L{ConnectionLost}.  This is weird; however, it is simply an attempt at 
> 253	 	        a faithful re-implementation of the behavior provided by 
> 254	 	        L{twisted.internet.ssl}. 
> ```
> 
> Why? AFAICT it's still true (I'll happily admit it's noisy, but it should probably remain documented).
> 

I found the comment a little confusing IIRC.  I'll fix it up (and document its interaction with the new handshake state stuff).

>  4. ```def notifyHandshakeDone(self):``` is missing a docstring

The docstring is in ISSLTransport.  Should I duplicate it?  Should I reference it?  I tried to follow existing practice, but I may have gotten it wrong.

>  5. ```def __tryHandshake(self):``` please use a single underscore for private methods

Will do.  There are some __ methods in Twisted -- is this a case of "follow the style of the file you're editing"?

>  6. What happens to the deferred previously returned from ```HandshakeCallbackContextFactory```?

It works exactly the same way as before.  It's arguably silly now, though -- should I try to get rid of it?

> 
> Thanks again, and I look forward to merging this patch :)

My pleasure :)

I'll submit an updated version soon (and probably change it to whenHandshakeDone -- I like that better).

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|amluto commented|
|-|-|

Replying to [amluto](#note_8):
> Replying to [lvh](#note_6):
> > Hi! Thanks for working on this, amluto!
> > 
> > A few nitpicks:
> > 
> >  1. Please document the types of the @ivars
> 
> Will do.

I think I had them mostly documented.  I improved _handshakeError.

> 
> >  2. Please don't put empty lines in between ivar defs
> 
> Will do.

I'm confused.  Where are these empty lines?  (Maybe I'm just being dense today.)

> 
> >  3. ```_handshakeDeferreds``` stops being a list when the handshake is done, but the ivar docs don't specify that
> 
> Will do.

Done.

> 
> >  4. The following thing is removed:
> > 
> > ```
> > 247	 	        <SNIP SNIP> This 
> > 248	 	        is used to control error reporting behavior.  If the handshake has not 
> > 249	 	        completed, the underlying L{OpenSSL.SSL.Error} will be passed to the 
> > 250	 	        application's C{connectionLost} method.  If it has completed, any 
> > 251	 	        unexpected L{OpenSSL.SSL.Error} will be turned into a 
> > 252	 	        L{ConnectionLost}.  This is weird; however, it is simply an attempt at 
> > 253	 	        a faithful re-implementation of the behavior provided by 
> > 254	 	        L{twisted.internet.ssl}. 
> > ```
> > 
> > Why? AFAICT it's still true (I'll happily admit it's noisy, but it should probably remain documented).
> > 
> 
> I found the comment a little confusing IIRC.  I'll fix it up (and document its interaction with the new handshake state stuff).

I may still be confused, but I don't think the comment was or is correct.  Can you elaborate?

> 
> >  4. ```def notifyHandshakeDone(self):``` is missing a docstring
> 
> The docstring is in ISSLTransport.  Should I duplicate it?  Should I reference it?  I tried to follow existing practice, but I may have gotten it wrong.
> 
> >  5. ```def __tryHandshake(self):``` please use a single underscore for private methods
> 
> Will do.  There are some __ methods in Twisted -- is this a case of "follow the style of the file you're editing"?

Done.

> 
> >  6. What happens to the deferred previously returned from ```HandshakeCallbackContextFactory```?
> 
> It works exactly the same way as before.  It's arguably silly now, though -- should I try to get rid of it?
> 
> > 
> > Thanks again, and I look forward to merging this patch :)
> 
> My pleasure :)
> 
> I'll submit an updated version soon (and probably change it to whenHandshakeDone -- I like that better).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to amluto|
|-|-|

Neither of the most recent two patches applies against trunk.

Please:

  1. Specify what your patches are against
  2. Generate svn-style patches (no "a/" and "b/" prefixes)
  3. Put everything in a single patch file

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|amluto commented|
|-|-|

tls_notify_v3.patch should apply against trunk.  Aside from rebasing it and rolling up the changes, I added a missing newline and moved the topfile into the correct place (I think).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_14"></a>|@habnabit set owner to @habnabit<br>@habnabit set status to `assigned`|
|-|-|

This patch contains no documentation outside of docstrings. There should be at least _some_ prose documentation for this. It should probably go in `doc/core/howto/ssl.xhtml`, preferably including a self-contained runnable example. I haven't tried writing code using this patch yet myself, and it would be nice to be able to run an example.

Additionally, adding methods to an interface breaks backward compatibility with anything that's implemented that interface externally. Instead, there should be a new interface for something which can notify on TLS handshake completion, and then make `TLSMemoryBIOProtocol` implement it.

Other than that, this generally looks good. I haven't done a thorough check through yet and won't mark the ticket as reviewed until I do. This is just preliminary.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|amluto commented|
|-|-|

I'll write up a patch to ```ssl.xhtml``` with a variant of the echo server.

Should the new interface be something general like IHandshakingTransport?

I think I'll wait to send more patches until this has a branch to minimize the amount of rebasing needed.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_16"></a>|@habnabit commented|
|-|-|

I'm terrible at naming. `ITLSHandshakingTransport` maybe? 

I'll make a branch later today when I have a chance to do a more thorough review.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|amluto commented|
|-|-|

If it were ```ITLSHandshakingTransport``` then it would be odd to reuse it for some other transport (SSH, SASL, QUIC, etc).  If it has a more general name, then any transport that goes through a connected-but-not-really-set-up phase could use it.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_18"></a>|@habnabit commented|
|-|-|

I considered that, and I _suppose_ the patch doesn't callback/errback with anything that's TLS-specific. So, okay, maybe this is generalized enough to be `IHandshakingTransport`. I'm fine with it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

> Additionally, adding methods to an interface breaks backward compatibility with anything that's implemented that interface externally.

This is entirely true but everyone else completely ignores this fact and adds things to interfaces all the time.  It's not fair to apply this criterion in a selective fashion against only certain contributions.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_19):
> > Additionally, adding methods to an interface breaks backward compatibility with anything that's implemented that interface externally.
> 
> This is entirely true but everyone else completely ignores this fact and adds things to interfaces all the time.  It's not fair to apply this criterion in a selective fashion against only certain contributions.

+1.

We should, by all means, have some more brainstorming sessions about how to address this consistently in the future, somewhere other than a ticket, though.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_21"></a>|@habnabit commented|
|-|-|

(In [40725]) Branching to 'tls-handshake-notification-6024'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_22"></a>|@habnabit commented|
|-|-|

(In [40726]) Applying patch from amulto. refs [#6024](https://github.com/twisted/twisted/issues/6024)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_23"></a>|@habnabit commented|
|-|-|

[Forced a build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/tls-handshake-notification-6024).

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_24"></a>|@habnabit set owner to amluto<br>@habnabit set status to `new`|
|-|-|

1. `whenHandshakeDone` in `tls.py` needs a docstring.
2. Is `_tlsShutdownFinished` always called? Even if the connection is lost midway through a handshake? It didn't look like there was a test for that scenario.
3. Could lines 490-507 of `tls.py` not be moved into `_handshakeSucceeded`? (After renaming it, of course.) It seems to be a bit of duplication.
4. Nit: I would say 'when the SSL handshake succeeds or fails' in the topfile, to make it more clear.
5. You have some [pyflakes](http://buildbot.twistedmatrix.com/builders/pyflakes/builds/1213/steps/pyflakes/logs/new%20pyflakes%20errors) and [twistedchecker](http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1511/steps/run-twistedchecker/logs/new%20twistedchecker%20errors) errors.

Thanks again for the contribution! I'm pretty excited for this to land. (Also sorry for misspelling your name in the commit message; I feel bad. ):

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_25"></a>|@habnabit commented|
|-|-|

Oh man I am just the worst at trac. Sorry about that. Let's try again:

 1. whenHandshakeDone in tls.py needs a docstring. 
 2. Is _tlsShutdownFinished always called? Even if the connection is lost midway through a handshake? It didn't look like there was a test for  that scenario.
 3. Could lines 490-507 of tls.py not be moved into _handshakeSucceeded? (After renaming it, of course.) It seems to be a bit of duplication.
 4. Nit: I would say 'when the SSL handshake succeeds or fails' in the topfile, to make it more clear.
 5. You have some  pyflakes and  twistedchecker errors.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_26"></a>|@habnabit commented|
|-|-|

I was also just reminded: use `L{}` for `True`, `False`, and `None`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

Thanks for the continued work on this issue.

> This is entirely true but everyone else completely ignores this fact and adds things to interfaces all the time. It's not fair to apply this criterion in a selective fashion against only certain contributions.

This comment (of mine) notwithstanding, if you're changing an interface then you *do* at least have to update the implementations of that interface in Twisted.

`twisted.internet._oldtls` is deprecated but hasn't actually been removed yet.  It would be very wrong to change the interface while leaving the old, now incomplete, TLS implementation in place.

The howto-style documentation is still missing.  I'm sure the plan of adding it before merging this change will be followed through but it's too bad documentation is only being added at such a late stage in the development process.  Actually trying to *use* and *explain* an API often reveals shortcomings of the API and suggests changes that should be made to it.  It's awfully convenient to learn about those things before going to all the trouble of implementing the API rather than afterwards.

For example, I strongly suspect this feature would better be exposed as a protocol method.  Something like this new optional protocol interface:

```
class ITLSProtocol(Interface):
    def handshakeCompleted():
        pass
```

However, it's difficult to compare these two APIs without some example uses.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_28"></a>|amluto commented|
|-|-|

Replying to [habnabit](#note_25):
>  1. whenHandshakeDone in tls.py needs a docstring.

The copy in the interface definition has a docstring.  What's the standard practice here?  Should I reference one docstring from the other?  Should I duplicate it?

>  2. Is _tlsShutdownFinished always called? Even if the connection is lost midway through a handshake? It didn't look like there was a test for  that scenario.

Connection failure mid-handshake presumably looks just like a failed handshake, but I haven't verified that.  I'll add a test.

>  3. Could lines 490-507 of tls.py not be moved into _handshakeSucceeded? (After renaming it, of course.) It seems to be a bit of duplication.

I could do something like ```def _handshakeDone(deferred_func, result)``` that would be passed either ```Defer.callback``` or ```Defer.errback```.  It would be shorted and less duplicated, but it would also be a bit odd.  I'll play with it.

>  4. Nit: I would say 'when the SSL handshake succeeds or fails' in the topfile, to make it more clear.
>  5. You have some  pyflakes and  twistedchecker errors.

I'll fix those.

Replying to [exarkun](#note_27):
> Thanks for the continued work on this issue.
> 
> > This is entirely true but everyone else completely ignores this fact and adds things to interfaces all the time. It's not fair to apply this criterion in a selective fashion against only certain contributions.
> 
> This comment (of mine) notwithstanding, if you're changing an interface then you *do* at least have to update the implementations of that interface in Twisted.
> 
> `twisted.internet._oldtls` is deprecated but hasn't actually been removed yet.  It would be very wrong to change the interface while leaving the old, now incomplete, TLS implementation in place.

Fair enough.  I assume that removing _oldtls at the same time is not an option.

> 
> The howto-style documentation is still missing.  I'm sure the plan of adding it before merging this change will be followed through but it's too bad documentation is only being added at such a late stage in the development process.  Actually trying to *use* and *explain* an API often reveals shortcomings of the API and suggests changes that should be made to it.  It's awfully convenient to learn about those things before going to all the trouble of implementing the API rather than afterwards.
> 
> For example, I strongly suspect this feature would better be exposed as a protocol method.  Something like this new optional protocol interface:
> 
> ```
> class ITLSProtocol(Interface):
>     def handshakeCompleted():
>         pass
> ```
> 
> However, it's difficult to compare these two APIs without some example uses.

Hmm.  The thing I don't like about that is that I don't see how a protocol would know whether to expect ```handshakeCompleted``` to be called.  For example, with the ```whenHandshakeDone``` approach, a protocol could run over plain TCP or TLS and only wait for a handshake when running over TLS.  Another downside: new code would fail silently if run on an old version of Twisted.

If we stick with ```whenHandshakeDone```, it might pay to add ```renenogiate``` at the same time to avoid more interface churn.  I suspect that ```renegotiate``` would be straightforward to implement on top of this code.

I'll write up some kind of example.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_29"></a>|amluto removed owner|
|-|-|

I attached more fixes.  This adds an example, fixes pyflakes and the relevant twistedchecker stuff (I think, although I haven't added docstrings to test cases).

The open issues are:

 1. I'm still using ```C{True```} instead of ```L{True```}.  All of ```tls.py``` does that.  I think that fixing it should be a separate change.
 1. I'm still extended ```ISSLTransport```.  I'd like to see some kind of consensus on what to do before changing that.  (The example I added should help.)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_30"></a>|@habnabit commented|
|-|-|

As a note, new examples should be using `twisted.internet.react` instead of importing the global reactor.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_31"></a>|@habnabit commented|
|-|-|

er, sorry, `twisted.internet.task.react`.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_32"></a>|@adiroiban commented|
|-|-|

Is the ``while True:`` code from ``_tryHandshake`` safe? I assume that for slow connections... or bad peers that handshake could take a while until in is done and this could block the thread.

I also wanted to know when handshake was done and I used the SSL.Context.set_info_callback method to hook a watcher and then SSL.Connection.set_app_data() to update the SSL.Connection state.

Maybe TLSMemoryBIOProtocol.makeConnection could set a deferred as app_data, and then the context info callback could call that deferred once handshake is done.

Does it make any sense?

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_33"></a>|amluto commented|
|-|-|

Replying to [adiroiban](#note_32):
> Is the ``while True:`` code from ``_tryHandshake`` safe? I assume that for slow connections... or bad peers that handshake could take a while until in is done and this could block the thread.

The call to do_handshake is nonblocking, so nothing blocks.  The only way the loop can keep looping is if pyOpenSSL keeps saying WantWriteError, which would mean that pyOpenSSL is continually writing handshake-related data.  This can't happen, barring a major bug in pyOpenSSL.

> 
> I also wanted to know when handshake was done and I used the SSL.Context.set_info_callback method to hook a watcher and then SSL.Connection.set_app_data() to update the SSL.Connection state.
> 
> Maybe TLSMemoryBIOProtocol.makeConnection could set a deferred as app_data, and then the context info callback could call that deferred once handshake is done.

The whole point of this patch is to do this cleanly instead of using hacks like that.

--Andy

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

(In [41791]) Apply tls_notify_6204_v4.patch

refs [#6024](https://github.com/twisted/twisted/issues/6024)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

(In [41793]) Branching to 'tls-handshake-notification-6024-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun set owner to amluto<br>@exarkun set status to `new`|
|-|-|

Thanks all.  Sorry about the delay in getting this reviewed. 

I made some changes, mostly boring coding standard and formatting things.  I noticed two things:

  * The `WantWriteError` exception handler in `_tryHandshake` appears to be untested.
  * shawnn's patch to make `whenHandshakeDone` work after `startTLS` seems like an important one (`startTLS` is supposed to make the transport provide `ISSLTransport` and `whenHandshakeDone` is now part of that interface...).  His patch looks quite simple but I have no idea if it is correct/complete - it needs unit tests as well. :)  I didn't apply his patch to the branch yet.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_38"></a>|amluto commented|
|-|-|

Replying to [exarkun](#note_37):
> Thanks all.  Sorry about the delay in getting this reviewed. 
> 
> I made some changes, mostly boring coding standard and formatting things.  I noticed two things:
> 
>   * The `WantWriteError` exception handler in `_tryHandshake` appears to be untested.

This may be rather difficult to test.  I suspect that the only way is to force the bio buffer size to be extremely small.  I'll fiddle with it.

>   * shawnn's patch to make `whenHandshakeDone` work after `startTLS` seems like an important one (`startTLS` is supposed to make the transport provide `ISSLTransport` and `whenHandshakeDone` is now part of that interface...).  His patch looks quite simple but I have no idea if it is correct/complete - it needs unit tests as well. :)  I didn't apply his patch to the branch yet.

Should I look at that patch?  It seems straightforward, but I'm a little vague on the whole starttls mechanism.

Is there any decision on what interface whenHandshakeDone should live in?

> 
> Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun commented|
|-|-|

> Is there any decision on what interface whenHandshakeDone should live in?

I'm aware of two options that have been considered.

First, the option that is currently implemented in the branch: dd the method to the existing interface (what the branch does).  Applications can discover if the feature is available using `getattr` or by checking which version of Twisted is being used.

Second, introduce a new interface that extends `ISSLTransport` with this new method (`ISecureTransport`?  Or will we get sued for that... ;).  Update the implementations in Twisted to declare they implement this new interface instead.  Deprecate the old interface.  Applications can discover if the feature is available by checking for the new interface instead of the old interface.

Third, play some tricks with interface versioning.  This isn't something we've attempted in Twisted before.  I have some untested ideas about adding a `version` `Attribute` to interfaces the first time we want to change them, requiring that its value be a monotonically increasing integer starting with 1 (version 0 is the version of the interface that is missing the attribute) and then documenting the rest of the interface in terms of that value (ie "this attribute was introduced in version 3 of the attribute").  Again, this involves breaking new ground and some amount of trial and error to figure out if the approach is workable at all, let alone whether we really like it or not.

By the way, you've reminded me of another potential issue.  `IReactorSSL` only uses `twisted.protocols.tls` if pyOpenSSL 0.10 or newer is available.  For older versions of pyOpenSSL it falls back to `twisted/internet/_oldtls.py`.  This new feature hasn't been implemented in `_oldtls.py` so the TLS transport applications get when pyOpenSSL is older than 0.10 will be missing this feature.  With the 2nd (or 3rd) option above this isn't necessarily a problem - at least `_oldtls.py` will still advertise that it does not implement this feature (by declaring that it only implements `ISSLTransport` and not the new interface).  With the first option above, `_oldtls.py` is now lying about its capabilities. :(

Personally, I prefer the 2nd option for basically this reason.  Changing an interface generally makes liars out of all existing code.  Even if we fix `_oldtls.py` so that it really implements this new feature that won't fix any third-party implementations of the interface.  There will be a window during which `if ISSLTransport.providedBy(transport): transport.whenHandshakeDone()` does *not* actually work reliably.  Depending on how active the development of third-party libraries implementing this interface is, the window could stretch out for quite a long time.

However, as I've mentioned elsewhere, the Twisted project doesn't actually have an official policy of not adding new attributes or methods to interfaces.  I would like it if it did but that's an issue to be resolved elsewhere, not on this ticket. :)

So I think the `ISSLTransport` change implemented in this branch (plus fixes for `_oldtls.py` (oh, and maybe `twisted.web.test.requesthelper.DummyChannel.SSL` - sorry)) are acceptable if that's they way you really want to go.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun commented|
|-|-|

> Should I look at that patch? It seems straightforward, but I'm a little vague on the whole starttls mechanism

If you decide to go with `ISSLTransport.whenHandshakeDone` then I think `startTLS` does need to be considered here, yes.

The purpose of `startTLS` is to upgrade the transport from something that does not provide `ISSLTransport` to something that does provide `ISSLTransport`.  If the result was an upgrade to something that provides `ISSLTransport`-but-not-whenHandshakeDone I think that would be quite unfortunate.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_41"></a>|@adiroiban commented|
|-|-|

> The call to do_handshake is nonblocking, so nothing blocks. The only way the loop can keep looping is if pyOpenSSL keeps saying WantWriteError, which would mean that pyOpenSSL is continually writing handshake-related data. This can't happen, barring a major bug in pyOpenSSL.

I will give this code a try and see how it behaves.

My problem is not that the loops runs forever, but that it can take some time for the actual handshake to finish.

From my understanding for server scenario is that after the SERVER-HELLO reponse, the server will wait for client to send the secret key.
For client scenario, the client first waits for SERVER-HELLO

I searched openSSL code but could not find the timeout for handshake operation. I guess it rely on TCP timeout.. so if client lags in sending the secret key, the handshake might lag before continuing. TCP default timeouts are quite big.

This is why I think that hooking into connection state via a callback is closer to general Twisted async model.

Sorry for the noise. I am not an SSL/TLS expert so maybe I am wrong.


>>   * The `WantWriteError` exception handler in `_tryHandshake` appears to be untested.

> This may be rather difficult to test.  I suspect that the only way is to
> force the bio buffer size to be extremely small.  I'll fiddle with it.

I got into this issue/error by closing the SSL connection without writing anything to it.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun commented|
|-|-|

> Sorry for the noise. I am not an SSL/TLS expert so maybe I am wrong.

Thanks for looking into this!  Twisted's TLS code can only benefit from more eyes. :)

> My problem is not that the loops runs forever, but that it can take some time for the actual handshake to finish.

I think I'm convinced this loop is okay.  Let me try to rephrase the response amluto gave in his comment above.

`do_handshake` can basically do one of two things (actually it could do both of them but let's ignore that).
  1. It can decide that it is "our" turn to talk in the handshake (for example, "we" are the client and it is the beginning of the connection, in TLS the client speaks first so in this case it is time to send the "ClientHello" message).  In this case what happens is that `do_handshake` "writes" some bytes to the BIO (since we use a "memory BIO" this really just means copying some bytes into a buffer that we're paying attention to somewhere else).  This write has two possible results.
    1. It could succeed completely.  At this point it is no longer "our" turn to talk.  Therefore OpenSSL will try to read some bytes.  This will fail with WantReadError ("error!  I want to read but I cannot.").  In this case we break out of the loop.
    1. Or it could fail because there isn't enough room in the buffer to do the write.  This causes `do_handshake` to fail with WantWriteError ("error!  I want to write but I cannot.").  If this happens then `_flushSendBIO` is called.  This is the method that takes things out of the "memory BIO" and puts them into another buffer (Twisted's own send buffer).  This makes more room in the "memory BIO".  The loop isn't broken out of this time, instead we call `do_handshake` again.  This time it should be able to do the write and we go back to the case immediately above (where we break out of the loop).
  1. It can decide it is not "our" turn to talk in the handshake.  In this case it will try to read some bytes from the "memory BIO".  Either there will be enough bytes there to read or not:
    1. If there are enough bytes then the handshake will proceed a little further.  This lets us proceed to the point where it's "our" turn to talk then we go back to the first case.
    1. If there are not enough bytes then `do_handshake` will raise "WantReadError".  This is like 1.1 above and we break out of the loop.

Notice that in all of the `WantReadError` cases - the cases where it is now time to wait for more bytes to arrive from the network - we immediately break out of the loop.  Therefore we don't spend that waiting time inside `_tryHandshake`, we spend it in the reactor and doing other useful work until something calls `dataReceived` on the protocol again.

Hope that helped.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_44"></a>|@glyph commented|
|-|-|

(In [41998]) Branching to tls-handshake-notification-6024-3.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_45"></a>|@glyph commented|
|-|-|

Okay!  So, at this point, I think that adiroiban's suggestion about `info_callback` is actually completely the right approach, and not a hack.

There are various reasons that it might not be optimal for the initial handshake (it's somewhat challenging to implement, for one thing), but there's another problem: _renegotiation_.

[Renegotiation is "secure" these days](https://tools.ietf.org/html/rfc5746), which I take to mean that the renegotiation packet in the TLS stream itself is now itself authenticated and encrypted so that you can't get told to renegotiate by an arbitrary man in the middle.  Nevertheless, renegotiation can still result in compromised security; your peer can change their security parameters from those you find acceptable to those you find unacceptable.  Fundamentally, OpenSSL is _the thing in the stack here which actually understands the TLS bytes_, and the "info callback" is the way that it tells us what's going on with those bytes.  A renegotiation can happen at the wire level without any observable reason to call `do_handshake` up at our protocol's level.

The problem, of course, is that Twisted users who already needed this functionality will already be using the `info_callback`, and [pyOpenSSL won't actually let us get at that callback to keep sending that feedback to them compatibly](https://github.com/pyca/pyopenssl/issues/90), so for a general solution to this problem we will need some way for the user to unambiguously indicate that they are _interested_ in this callback and that we should manipulate the context for them, or perhaps put it in the context itself.

I'm already using the `info_callback` for [#5190](https://github.com/twisted/twisted/issues/5190) because I couldn't reliably abort the handshake in the right place (post-verification but pre-any-application-bytes-getting-through) from the "outside", so I think we should abandon the implementation in the branch.

In the process of thrashing around and discovering this, I did implement a significant number of useful test refactorings, and those should probably be applied to at least one other ticket.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_46"></a>|@glyph commented|
|-|-|

Oops, ambiguous antecedent alert: when I said "it" might not be optimal for the initial handshake, I meant the strategy implemented in the branch.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_47"></a>|amluto commented|
|-|-|

Replying to [glyph](#note_45):
> Okay!  So, at this point, I think that adiroiban's suggestion about `info_callback` is actually completely the right approach, and not a hack.
> 
> There are various reasons that it might not be optimal for the initial handshake (it's somewhat challenging to implement, for one thing), but there's another problem: _renegotiation_.
> 
> [Renegotiation is "secure" these days](https://tools.ietf.org/html/rfc5746), which I take to mean that the renegotiation packet in the TLS stream itself is now itself authenticated and encrypted so that you can't get told to renegotiate by an arbitrary man in the middle.  Nevertheless, renegotiation can still result in compromised security; your peer can change their security parameters from those you find acceptable to those you find unacceptable.  Fundamentally, OpenSSL is _the thing in the stack here which actually understands the TLS bytes_, and the "info callback" is the way that it tells us what's going on with those bytes.  A renegotiation can happen at the wire level without any observable reason to call `do_handshake` up at our protocol's level.

Your peer can always break the security of your connection by leaking the secret.  There's nothing you, Twisted, or anything else can do about that.

The renegotiation attacks were based on starting with a connection that shouldn't be trusted and renegotiating it to turn it into something else.  The key is that, in all those attacks, the initial connection was to the attacker.

I've never heard of an attack that allows an arbitrary MITM to force a renegotiation.

[...]

> 
> I'm already using the `info_callback` for [#5190](https://github.com/twisted/twisted/issues/5190) because I couldn't reliably abort the handshake in the right place (post-verification but pre-any-application-bytes-getting-through) from the "outside", so I think we should abandon the implementation in the branch.
> 

Why not?

If you want to prevent application bytes from being sent, don't send them until the handshake is complete.

If you want to prevent application bytes from being received, just buffer them until the handshake is complete.  I can probably guarantee that no data will be received until the handshake reports itself complete, so this should be unnecessary.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_48"></a>|@glyph commented|
|-|-|

Replying to [amluto](#note_47):
> Replying to [glyph](#note_45):

> Your peer can always break the security of your connection by leaking the secret.  There's nothing you, Twisted, or anything else can do about that.

And yet, browsers still display warnings when "mixed content" is present, or [simply block it entirely](https://support.google.com/chrome/answer/1342714?hl=en).  There are more threats on the internet, Horatio, than are dreamt of in your mathematical trust model :-).

Consider attacks that are partially on the node itself.  Maybe it's not totally pwned but the attacker can somehow convince its TLS layer to renegotiate down to the null cipher or something equivalent to it.  The point is, if the application has reason to verify anything about the post-handshake state of the connection other than the certificate, it has might have reason to examine subsequent handshakes as well as the first one.

> The renegotiation attacks were based on starting with a connection that shouldn't be trusted and renegotiating it to turn it into something else.  The key is that, in all those attacks, the initial connection was to the attacker.

Yes, sorry, what I was trying to say with my comments about renegotiation attacks was that I was _not_ specifically trying to defend against any existing, known attack, for which there are protocol-level fixes anyway.

> I've never heard of an attack that allows an arbitrary MITM to force a renegotiation.

Neither have I.

> > 
> > I'm already using the `info_callback` for [#5190](https://github.com/twisted/twisted/issues/5190) because I couldn't reliably abort the handshake in the right place (post-verification but pre-any-application-bytes-getting-through) from the "outside", so I think we should abandon the implementation in the branch.
> > 
> 
> Why not?

Well, I tried, and it was complicated and hard.  Then I tried using `info_callback`, and it was simple and easy.  So I'm going with that one.

> If you want to prevent application bytes from being sent, don't send them until the handshake is complete.

How, exactly, do I tell that the handshake is complete?  I know that `do_handshake` returned successfully, but I don't actually know 

> If you want to prevent application bytes from being received, just buffer them until the handshake is complete.  I can probably guarantee that no data will be received until the handshake reports itself complete, so this should be unnecessary.

What do you mean you can "guarantee" it?  Are you an OpenSSL developer, or are you saying you'll be contractually liable for any damages that Twisted sustains as a result of handshakes being implicitly terminated by `SSL_read` (as they are wont to do)?

Basically the problem here is that by doing all this stuff with `do_handshake`, we're replicating a state machine that already exists inside `OpenSSL`, and we can only do it partially.  Using `info_callback` is instead quite explicit, and works well.

I'm already doing this in [#5190](https://github.com/twisted/twisted/issues/5190) quite effectively for its use-case of verifying hostnames.  The only problem right now with doing this generally for notifying protocols with an arbitrary `contextFactory`, as we've [previously discussed on the pyOpenSSL tracker](https://github.com/pyca/pyopenssl/issues/90), is that existing applications which depend on `set_info_callback` will break.  Since that has been [the only advice from Twisted developers for how to do this fairly critical task in the meanwhile](https://stackoverflow.com/questions/4954255/how-to-check-that-tls-handshake-was-finalize-in-twisted/4958019#4958019), it seems likely that there's at least one application floating around out there that we should try not to break.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_49"></a>|@glyph commented|
|-|-|

Replying to [glyph](#note_48):

> How, exactly, do I tell that the handshake is complete?  I know that `do_handshake` returned successfully, but I don't actually know 

Oh, trac, one day we'll make it possible to edit comments.  Or throw you in the garbage.

As I was saying, I know that `do_handshake` completed successfully, but I don't know exactly what state it's in.  According to OpenSSL, the way to determine the exact state of the phases of the handshake is ... wait for it ... the info callback ;-).

Sorry about the truncation,

-glyph

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_50"></a>|@exarkun commented|
|-|-|

> According to OpenSSL, the way to determine the exact state of the phases of the handshake is ... wait for it ... the info callback

Are the phases other than "the handshake is done" important?

```
The following return values can occur:

  1.  The TLS/SSL handshake was successfully completed, a TLS/SSL connection has been established.
```

Or are you referring to the fact that renegotiation is effectively restarting the handshake and the return value of `SSL_do_handshake` is useless in observing when this happens?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_51"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_50):
> Or are you referring to the fact that renegotiation is effectively restarting the handshake and the return value of `SSL_do_handshake` is useless in observing when this happens?

Exactly.  What I meant was: `do_handshake` will totally reliably tell you that the handshake completed, but how do you know you need to call it?

We already have working code that deals with handshake-in-progress error codes from arbitrary read and write calls; we could change that to have a bunch of special-case logic to keep calling `do_handshake` for the initial handshake, but, why? It doesn't seem like that lets us do anything except avoid the info callback, and there doesn't seem to be any good reason to avoid it.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_52"></a>|amluto commented|
|-|-|

Replying to [glyph](#note_51):
> Replying to [exarkun](#note_50):
> > Or are you referring to the fact that renegotiation is effectively restarting the handshake and the return value of `SSL_do_handshake` is useless in observing when this happens?
> 
> Exactly.  What I meant was: `do_handshake` will totally reliably tell you that the handshake completed, but how do you know you need to call it?
> 
> We already have working code that deals with handshake-in-progress error codes from arbitrary read and write calls; we could change that to have a bunch of special-case logic to keep calling `do_handshake` for the initial handshake, but, why?

That's exactly what this branch does.  And it's not exactly special-case; it's just doing what the OpenSSL docs say to do.

> It doesn't seem like that lets us do anything except avoid the info callback, and there doesn't seem to be any good reason to avoid it.

It lets us have a decent interface for determining when the handshake is done.  The info callback is ugly, exposes OpenSSL internals, and, worst of all, it's a callback.  That means that it's liable to be re-enter Python code at inopportune moments, such as when someone calls write.

Usually a handshake will complete as a result of a call from the reactor, but there's no guarantee that this is a case.  Something could race against the reactor, call transport.write, and cause the handshake to finish.  That will cause the info callback to fire.  Ouch.

This is unlikely.  It may even be impossible, but I doubt that.  I can certainly see it causing hard-to-predict breakage.

This branch shouldn't do that.  I'll take a look (probably in a week or two -- I'm currently swamped with other things) to make sure that it can't happen.  But at least by using do_handshake, we control when things happen.


* It's late.  I could have some of the names wrong here.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_53"></a>|@glyph commented|
|-|-|

Replying to [amluto](#note_52):
> Replying to [glyph](#note_51):
> > Replying to [exarkun](#note_50):
> > > Or are you referring to the fact that renegotiation is effectively restarting the handshake and the return value of `SSL_do_handshake` is useless in observing when this happens?
> > 
> > Exactly.  What I meant was: `do_handshake` will totally reliably tell you that the handshake completed, but how do you know you need to call it?
> > 
> > We already have working code that deals with handshake-in-progress error codes from arbitrary read and write calls; we could change that to have a bunch of special-case logic to keep calling `do_handshake` for the initial handshake, but, why?
> 
> That's exactly what this branch does.

Except the branch doesn't quite work.  Specifically, it doesn't seem to discard unauthenticated bytes that were written by a `_tlsConnection.send` or read by a `_tlsConnection.recv`.  I know that in principle this is totally _possible_, but it's not working yet, and starting over with the much simpler `info_callback` just fixed the issue with far less of an impact (almost none, in fact) on the structure of `twisted.protocols.tls`.

> And it's not exactly special-case; it's just doing what the OpenSSL docs say to do.

This is an appeal to authority; and not a particularly good authority at that.  The OpenSSL docs generally think a long list of functions are an adequate replacement for a tutorial, and they also mostly assume you're doing synchronous I/O.

> > It doesn't seem like that lets us do anything except avoid the info callback, and there doesn't seem to be any good reason to avoid it.
> 
> It lets us have a decent interface for determining when the handshake is done.

This is begging the question.  "Why is the info callback not good?  Because the thing we would get if we used the info callback is bad."

> The info callback is ugly, 

Again: not a real argument.  Details, please.

> exposes OpenSSL internals,

With "public" APIs like `int SSL_get_ex_new_index(long argl, char *argp, int (*new_func);(void), int (*dup_func)(void), void (*free_func)(void))` it's not clear to me that OpenSSL has anything _but_ internals.  We would be kidding ourselves to say that any of this code would be re-usable to other TLS implementations.  (Most other TLS APIs are so much easier to work with, though, that there's no particular reason to _want_ to share this code.)

> and, worst of all, it's a callback.  That means that it's liable to be re-enter Python code at inopportune moments, such as when someone calls write.

Most things in Twisted are callbacks.  If we need to break out of it in order to avoid reentrant application code, then that's fine; we can easily enqueue whatever needs to be done into a call queue on the TLS protocol wrapper.  But the work I've done so far in [#5190](https://github.com/twisted/twisted/issues/5190) indicates that we probably don't need to worry about that.

> Usually a handshake will complete as a result of a call from the reactor, but there's no guarantee that this is a case.  Something could race against the reactor, call transport.write, and cause the handshake to finish.  That will cause the info callback to fire.  Ouch.

Uh... no?  You're not allowed to diddle these objects from other threads.  If you do, and it breaks, you can keep both pieces.  The reactor is fully in control of all the I/O going into these things.  What exactly do you imagine racing with it?

> This is unlikely.  It may even be impossible, but I doubt that.  I can certainly see it causing hard-to-predict breakage.

I literally can't imagine any valid API usage that would cause it, so I honestly don't know what you're talking about.

> This branch shouldn't do that.  I'll take a look (probably in a week or two -- I'm currently swamped with other things) to make sure that it can't happen.  But at least by using do_handshake, we control when things happen.

I'm sorry to say that in a week or two, this will all probably be finished and moot :-).  (Although I've said similar things about 6 years ago on tickets that are still open, so...)

Nevertheless, if there's a good reason to re-factor the implementation to not use the info callback - and I'm still waiting to hear a _good_ reason - it should definitely be possible to re-factor the internals to call `do_handshake` rather than depend on the info callback and provide the same interface to applications, and I'm happy for you to do that.

**Comment by twisted-trac** [comment]
|<img alt="amluto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_54"></a>|amluto commented|
|-|-|

Replying to [glyph](#note_53):
> Replying to [amluto](#note_52):
> > Replying to [glyph](#note_51):
> > > Replying to [exarkun](#note_50):
> > > > Or are you referring to the fact that renegotiation is effectively restarting the handshake and the return value of `SSL_do_handshake` is useless in observing when this happens?
> > > 
> > > Exactly.  What I meant was: `do_handshake` will totally reliably tell you that the handshake completed, but how do you know you need to call it?
> > > 
> > > We already have working code that deals with handshake-in-progress error codes from arbitrary read and write calls; we could change that to have a bunch of special-case logic to keep calling `do_handshake` for the initial handshake, but, why?
> > 
> > That's exactly what this branch does.
> 
> Except the branch doesn't quite work.  Specifically, it doesn't seem to discard unauthenticated bytes that were written by a `_tlsConnection.send` or read by a `_tlsConnection.recv`.  I know that in principle this is totally _possible_, but it's not working yet, and starting over with the much simpler `info_callback` just fixed the issue with far less of an impact (almost none, in fact) on the structure of `twisted.protocols.tls`.

What call to _tlsConnection.recv?  There's only one call site, and it's guarded by a check that the handshake is done?

And what call to _tlsConnection.send?  Can you explain how an application that waits for the handshake to complete before calling transport.write can possibly call _tlsConnection.send before the handshake is done?

> > The info callback is ugly, 
> 
> Again: not a real argument.  Details, please.

I'm done with this argument.  As far as I know, the only real issue with this branch is that the interface needs a bit of work.  If you can solve the same problem with less code or cleaner code using info_callback, be my guest.

> > Usually a handshake will complete as a result of a call from the reactor, but there's no guarantee that this is a case.  Something could race against the reactor, call transport.write, and cause the handshake to finish.  That will cause the info callback to fire.  Ouch.
> 
> Uh... no?  You're not allowed to diddle these objects from other threads.  If you do, and it breaks, you can keep both pieces.  The reactor is fully in control of all the I/O going into these things.  What exactly do you imagine racing with it?

This has nothing whatsoever to do with threads.  Application code could call transport.write in response to an event on a different socket.  That being said, I'm not immediately seeing how the particular issue can happen.

> 
> Nevertheless, if there's a good reason to re-factor the implementation to not use the info callback - and I'm still waiting to hear a _good_ reason - it should definitely be possible to re-factor the internals to call `do_handshake` rather than depend on the info callback and provide the same interface to applications, and I'm happy for you to do that.

I *already did that*.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_55"></a>|@Lukasa removed owner|
|-|-|

Because I make bad life decisions, I've taken a swing at this using glyph's suggested approach of the info callback. It turns out to not be that much code. You can find the diff on [GitHub in pull request [#7313](https://github.com/twisted/twisted/issues/7313)](https://github.com/twisted/twisted/pull/189).

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_56"></a>|@Lukasa commented|
|-|-|

In changeset 231bd6bf5aa726021c19cd75e5ef081309175854
```
#!CommitTicketReference repository="" revision="231bd6bf5aa726021c19cd75e5ef081309175854"
Add topfile for #6024.
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_58"></a>|@glyph set owner to @Lukasa|
|-|-|

Hooray for bad life decisions!

I've given this a [run through the buildbots](https://buildbot.twistedmatrix.com/boxes-supported?branch=handshake-6024) so you can address any test failures, doc or style issues that crop up there; I won't enumerate them :).

I do have a few concerns that I think need to be addressed:

  1. The most significant concern here is a compatibility concern, which is ironically made clear by your change to the tests.  Given [https://github.com/pyca/pyopenssl/issues/90 the lack of `get_info_callback`], we don't have a way to keep track of ''existing ContextFactory objects that use the `info_callback` for their own ends''.  This is why, when we started using the `info_callback` client-side, we went to some trouble to ensure that you would never get the new behavior unless you touched the new API: if we used it unconditionally, then others who were employing the same strategy to work around Twisted's lack of such functionality would find their applications silently broken.  In this case, the breakage wouldn't be quite as silent as in the verification case, but nevertheless it would lead to a rocky transition: if I'm using the info_callback to get handshake completion notification right now, that will break when I upgrade Twisted, and then I have to add a version-conditional if I want to support multiple Twisteds.  The way to address this, I think, is to move the implementation-detection logic from the place where `handshakeCompleted` is _called_ to where the `info_callback` itself is _added_.  In `optionsForClientTLS`, of course, we already set an `info_callback`, so there's no compatibility concern there.  But in the other case (pretty much all server-side cases, plus older clients using custom `ContextFactory` implementations), we might have an `info_callback` that is critical to the application that we can't see.  However, said application wouldn't have been able to import or use `IHandshakeListener` in older versions of Twisted, so by providing that interface you are declaring interest in the info_callback being added.
  2. There's also the issue of re-entrancy, which both adiroiban and amluto have raised.  We definitely need some test coverage for doing `self.transport.write()` inside `handshakeCompleted`; it's not even an edge-case really, it's the main thing you might want to do.  I do not like the idea of establishing a dependency on the reactor.  We _definitely_ don't need `callLater(0.5)`; at most we'd need a `callLater(0)`.  The real reentrancy problem is `_tlsConnection.send` calling `_tlsConnection.send` again; the kick-off `do_handshake` will never get to this step of the handshake, and `dataReceived` calling `dataReceived` is a bug. So `TLSMemoryBIOProtocol._write` should guard against this kind of reentrancy by making reentrant calls append to `_appSendBuffer` .  If testing indicates the need for that, of course - but adiroiban's reports of infinite recursion suggest that it will :).
  3. `handshakeCompleted` on `TLSMemoryBIOProtocol` is unnecessary public API.  The other interfaces relayed between the 'low' level of TCP and the 'high' level of the protocol listening over TLS are there because the various layers may need to reach between each other.  But the TCP layer never needs to reach through to call `handshakeCompleted`; that is `TLSMemoryBIOProtocol`'s own job; the `info_callback` can just call straight through to the private attribute since it's within the same... conceptual namespace, I guess, as a method on `TLSMemoryBIOProtocol` itself.

Honestly after reviewing this code, and with the benefit of an additional 2+ years of experience with both TLS and the OpenSSL API, amluto's original comments make a lot more sense to me, and I'm less sure that the info callback is a good idea.  I wish he'd stuck around to explain or resumbit originally, but I can see why my comments didn't make much sense to him, either.  So let me try to answer my own question: why isn't the info callback good?

  1. At least in the absence of `get_info_callback`, it's not cooperative, so Twisted has to either own it, or not.  This introduces the relatively complex compatibility concern above.
  2. To expand on what they said about it being "a callback": it's called ponentially reentrantly _inside numerous places in OpenSSL_, potentially in `Connection.do_handshake`, `Connection.bio_write`, and maybe `Connection.send`.  It's not entirely clear when, and so guards against reentrancy have to be carefully written around all of them.  Except not really because we _do_ know where reentrancy is a risk, because we know stuff about where the handshake might succeed.  But if we're encoding that information in the implementation anyway - why not just call `do_handshake`?
  3. The point about private APIs has more credibility now, as OpenSSL has been aggressively distinguishing between "private" and "public" APIs.

The major argument against using `do_handshake`, as amluto's patch does, appears to be renegotiation.  Given all the problems that there have been with renegotiation in TLS, and its dubious utility, I went to have a look and see if maybe we could just disable renegotiation entirely.  Unfortunately this doesn't seem to be possible.

But, we _already have_ a flag called `_handshakeDone` in `TLSMemoryBIOProtocol`, which doesn't care about renegotiation at all; so this rather puts the lie to [#comment:48 my argument about "replicating a state machine that already exists inside OpenSSL"].  Clearly we have to replicate this state machine, we're already doing it in ad-hoc way, and what we care about and are testing in this change is _just the initial handshake_.  If we are going to support renegotiation notifications at all, I think we should actually have a _separate_ interface, which gives you both the old security parameters and the new ones.  `handshakeCompleted` is really intended for applications to be able to register interest in the beginning of the secure connection, not in random details of cipher suites changing later; renegotiation is a very sophisticated TLS power-user feature, and mixing them together by invoking the same callback again is just giving attackers a way to confuse otherwise-correct applications by causing them to execute a seldom-seen code path that you can't even provoke with twisted by itself (we don't expose `Connection.renegotiate` after all).

So after thinking about this way too hard and way too long, I think we _should_ call `do_handshake` until the handshake is done, and watch its return value to call this callback.  Nevertheless, thanks for doing the implementation, Lukasa, because I don't know that I would have been able to fully understand this without seeing both.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_59"></a>|@glyph removed owner|
|-|-|

OK, I took my own crack at it.  This is sort of a hybrid of the two approaches - amluto's implementation strategy and lukasa's interface.  https://github.com/twisted/twisted/pull/191

In the meanwhile, I tried to simplify the implementation as much as I could, and make it easier to read; I killed the duplicative `_writeBlockedOnRead` attribute and the totally undocumented/unnecessary `_shuttingDown`; the buffering and unbuffering of outgoing writes now reside in their own methods, which should make it a bit more straightforward what is going on.

And, of course: no usage of the info callback.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_60"></a>|@Lukasa commented|
|-|-|

Alright, so after some review and discussion I feel like I'm happy with this patch. This looks like it does the appropriate dance for handshaking, which is neat. I'd like to see the tests in `test_tls.py`, which currently rely on setting an info callback, changed to use this notification mechanism (in part because it'd validate that it works!). However, it's reasonable to consider that change part of a separate patch. (Though I made the appropriate changes in my branch, so it should be fairly easy to just pull them across.)

Delaying review acceptance until the builders are done and we know what's going on with coverage.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_61"></a>|@Lukasa set owner to @glyph|
|-|-|

Ok glyph, one small bit of coverage notes: `_unbufferPendingWrites` doesn't have full coverage: in particular, there's no coverage of the unbuffer/rebuffer code section here https://codecov.io/gh/twisted/twisted/pull/191/compare#747769737465642F70726F746F636F6C732F746C732E7079-585

Writing this test is going to be a bit annoying, but it would be really good to confirm that it actually works. I'd like to see that addressed.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_62"></a>|@Lukasa commented|
|-|-|

Also pyflakes: https://buildbot.twistedmatrix.com/builders/pyflakes/builds/267/steps/pyflakes/logs/new%20pyflakes%20errors

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_63"></a>|@Lukasa commented|
|-|-|

A note: this blocks [#8489](https://github.com/twisted/twisted/issues/8489).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_64"></a>|@glyph set owner to @Lukasa|
|-|-|

I've fixed the pyflakes issues, and covered all the lines in the patch.  You were absolutely right: writing that test _was_ super annoying!  I'd love it if you could give it a once-over before I hit the 'merge' button, though; mainly have a look at https://github.com/twisted/twisted/pull/191/commits/f9b770669b4d12fe9432eb5b17070134cd74150b and see if the explanation makes sense to you.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_65"></a>|@Lukasa set owner to @glyph|
|-|-|

Ok, the new commits look good. I'd like some understanding of what you plan to do with the new invalid reference to Error, but given that this is the result of a weird interaction with pydoctor I don't necessarily want us to sit on this ticket until we understand what's going on here. Feel free to create a new ticket for that, and then go ahead and merge this when you feel ready to.

Nicely done all, I'm glad we got here eventually!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_66"></a>|GitHub set status to `closed`|
|-|-|

In changeset e309d6309504ebb859e77d6c5ee3df6144224dc4
```
#!CommitTicketReference repository="" revision="e309d6309504ebb859e77d6c5ee3df6144224dc4"
Merge pull request #191 from twisted/handshake-6024-2

Author: glyph, amluto

Reviewer: lukasa, habnabit, exarkun, glyph

Fixes: #6024

The new interface IHandshakeListener that can be implemented by any Protocol provides a callback that is called when the TLS handshake has been completed, allowing Protocols to make decisions about the TLS configuration before application data is sent.
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_67"></a>|@glyph commented|
|-|-|

Replying to [Lukasa](#note_65):
> Ok, the new commits look good. I'd like some understanding of what you plan to do with the new invalid reference to Error, but given that this is the result of a weird interaction with pydoctor I don't necessarily want us to sit on this ticket until we understand what's going on here. Feel free to create a new ticket for that, and then go ahead and merge this when you feel ready to.

The invalid ref is a tooling issue, probably an intermittent intersphinx thing, so I definitely did not want to block on it.  I think we may need to have some better way of dealing with intersphinx inventory caching so that downtime on 3rd-party sites doesn't cause build failures.  In any case, the source is correctly expressing the relationship, so we can deal with it separately.

> Nicely done all, I'm glad we got here eventually!

Quite so!
